src/compiler/checker.ts

```diff
@@ -27929,10 +27929,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     if (isMatchingConstructorReference(right)) {
                         return narrowTypeByConstructor(type, operator, left, assumeTrue);
                     }
-                    if (isBooleanLiteral(right)) {
+                    if (isBooleanLiteral(right) && !isAccessExpression(left)) {

```

Comment:
It turns out that the original `type` can be reassigned based on `optionalChainContainsReference`. That messed up the else branch because `narrowTypeByBooleanComparison` was operating on the wrong type.

I've also tried keeping `originalType` around and pass that to `narrowTypeByBooleanComparison`. That fixed the reported issue~ related to the wrong type in the else branch but it broke the same type within the if block. This happened because originally the intention was for that reassigned `type` to be returned through the `return` at the bottom of this function but I changed it to return `originalType` through `narrowTypeByBooleanComparison`.

So all in all, this seems like the simplest fix and reflects the intention the best. The motivation behind the original change was to improve non-access expression narrowing. Access expressions are already handled by discrimination-based narrowing etc. 

Even though it worked previously (as far as I know) for non-optional chain access expressions, this helps them avoid some redundant work too.

Reply:
This seems fine as a quick fix for now, though I suspect that we could try and do better later.

Reply:
Then again, not sure what "better" even means; an expression like `(options?.a === false || options.b` is not the same as `(!options?.a || options.b)` so realistically I don't think there's any extra info gained here.